akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,403 @@
+        event.getClass.getName, seqNr, persistenceId)
+
+      // FIXME no callback for users?
+
+      Behaviors.stopped
+    }
+
+  }
+
+  // the active phase switches between PersistingEvents and HandlingCommands;
+  // we do this via a var instead of behaviours to keep allocations down as this will be flip/flaping on every Persist effect
+  private[this] var phase: EventsourcedRunningPhase = HandlingCommands
+
+  override def onMessage(msg: Any): Behavior[Any] = {
+    log.info("RUNNING onMessage: " + msg + s" @ ${phase.name}")
+    val res = msg match {
+      // TODO explore crazy hashcode hack to make this match quicker...?
+      case SnapshotterResponse(r) ⇒ onSnapshotterResponse(r)
+      case JournalResponse(r)     ⇒ phase.onJournalResponse(r)
+      case command ⇒
+        // the above type-check does nothing, since Command is tun
+        command match {
+          case IncomingCommand(wrapped: Command @unchecked) ⇒ // FIXME is this used? (I think not)
+            log.info("Unwrapping IncomingCommand...") // FIXME do we really need this?
+            phase.onCommand(wrapped)
+
+          case UnstashedCommand(wrapped: Command @unchecked) ⇒ // FIXME do we really need this? seems no
+            log.info(s"Applying unstashed command: ${wrapped}")
+            phase.onCommand(wrapped)
+
+          case _ ⇒
+            // we cast explicitly to fail early in case of type mismatch
+            val c = command.asInstanceOf[Command]
+            phase.onCommand(c)
+        }
+    }
+    println(s"res = ${res} +++ ${context.self}")
+    res
+  }
+
+  // ----------
+
+  def applySideEffects(effects: immutable.Seq[ChainableEffect[_, S]]): Behavior[Any] = {
+    var res: Behavior[Any] = same
+    val it = effects.iterator
+
+    // if at least one effect results in a `stop`, we need to stop
+    // manual loop implementation to avoid allocations and multiple scans
+    while (it.hasNext) {
+      val effect = it.next()
+      applySideEffect(effect) match {
+        case _: StoppedBehavior[_] ⇒ res = Behaviors.stopped
+        case _                     ⇒ // nothing to do
+      }
+    }
+
+    res
+  }
+
+  def applySideEffect(effect: ChainableEffect[_, S]): Behavior[Any] = effect match {
+    case _: Stop.type @unchecked ⇒
+      Behaviors.stopped
+
+    case SideEffect(callbacks) ⇒
+      callbacks(state)
+      same
+  }
+
+  def applyEvent(s: S, event: E): S =
+    eventHandler(s, event)
+
+  @tailrec private def applyEffects(msg: Any, effect: Effect[E, S], sideEffects: immutable.Seq[ChainableEffect[_, S]] = Nil): Behavior[Any] = {
+    log.info(s"ApplyEffects: ${msg} => Effect: ${effect}")
+
+    effect match {
+      case CompositeEffect(Some(persist), currentSideEffects) ⇒
+        applyEffects(msg, persist, currentSideEffects ++ sideEffects) // "unwrap" & recur
+
+      case CompositeEffect(_, currentSideEffects) ⇒
+        applySideEffects(currentSideEffects ++ sideEffects) unlessStopped tryUnstash(context, this)
```

Comment:
This is triggered by `onCommand`, which is triggered by `onMessage` so stopped should be fine here as well

Reply:
Works here, indeed